File: frontend/src/features/message/http/api/GetConversationHistory.ts
import type { MessageConversationId } from '@/features/message/model/MessageConversationId'
import type { MessageHistoryResponse } from '@/features/message/http/response/MessageHistoryResponse'
import type { MessageId } from '@/features/message/model/MessageId'
import { messageConversationIdValue, messageIdValue } from '@/features/message/lib/message-parsers'
import { fromMessageHistoryResponse } from '@/features/message/http/codec'
import { requestJson } from '@/shared/api/http-client'

export function getConversationHistory(
  conversationId: MessageConversationId,
  options: { beforeMessageId?: MessageId | null; limit?: number } = {},
): Promise<MessageHistoryResponse> {
  const url = new URL(`/api/messages/conversations/${encodeURIComponent(messageConversationIdValue(conversationId))}/messages`, window.location.origin)
  if (options.beforeMessageId) {
    url.searchParams.set('before', messageIdValue(options.beforeMessageId))
  }
  if (options.limit) {
    url.searchParams.set('limit', String(options.limit))
  }

  return requestJson(url.pathname + url.search, fromMessageHistoryResponse)
}

File: backend/src/main/scala/domains/message/http/api/GetConversationHistory.scala
package domains.message.http.api

import cats.effect.IO
import domains.auth.http.AuthenticatedApi
import domains.auth.objects.AuthUser
import domains.message.http.codec.MessageHttpCodecs.given
import domains.message.objects.response.MessageHistoryResponse
import domains.message.objects.{MessageConversationId, MessageId}
import domains.message.table.message.{DirectMessageTable, MessageConversationTable}
import io.circe.Encoder
import org.http4s.{Method, Request, Status}
import shared.http.{ApiMessages, ApiPath, HttpApiError, PathParams}

import java.sql.Connection

object GetConversationHistory extends AuthenticatedApi[GetConversationHistory.Input, MessageHistoryResponse]:

  private val defaultHistoryLimit = 50

  final case class Input(
    conversationId: MessageConversationId,
    beforeMessageId: Option[MessageId],
    limit: Option[Int]
  )

  override val method: Method = Method.GET
  override val path: ApiPath = ApiPath("/api/messages/conversations/:conversationId/messages")
  override val successStatus: Status = Status.Ok
  override protected val outputEncoder: Encoder[MessageHistoryResponse] = summon[Encoder[MessageHistoryResponse]]

  override def decode(request: Request[IO], pathParams: PathParams): IO[Input] =
    HttpApiError.fromEitherBadRequest {
      pathParams.require("conversationId").flatMap(MessageConversationId.parse).map { conversationId =>
        Input(
          conversationId = conversationId,
          beforeMessageId = request.uri.query.params.get("before").flatMap(rawId => MessageId.parse(rawId).toOption),
          limit = request.uri.query.params.get("limit").flatMap(_.toIntOption)
        )
      }
    }

  override def plan(
    connection: Connection,
    actor: AuthUser,
    input: Input
  ): IO[MessageHistoryResponse] =
    for
      maybeConversation <- MessageConversationTable.findConversationSummaryForUser(connection, actor.username, input.conversationId)
      conversation <- maybeConversation match
        case Some(conversation) => IO.pure(conversation)
        case None => HttpApiError.raise(HttpApiError.notFound(ApiMessages.directMessageConversationNotFound))
      messagesAndMore <- DirectMessageTable.listConversationMessages(
        connection,
        input.conversationId,
        input.beforeMessageId,
        input.limit.filter(_ > 0).getOrElse(defaultHistoryLimit)
      )
      facts <- DirectMessageTable.getConversationMessageFacts(connection, input.conversationId, actor.username)
      (messages, hasMore) = messagesAndMore
    yield MessageHistoryResponse(
      conversation = conversation,
      messages = messages,
      hasMore = hasMore,
      facts = facts
    )
